Mathlib/RingTheory/Binomial.lean

```diff
@@ -0,0 +1,120 @@
+/-
+Copyright (c) 2023 Scott Carnahan. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Scott Carnahan
+-/
+
+import Mathlib.Data.Rat.NNRat
+import Mathlib.RingTheory.Polynomial.Pochhammer
+
+/-!
+# Binomial semirings and binomial rings
+
+In this file we introduce binomial semirings and binomial rings, together with related functions
+like binomial coefficients.
+
+According to our main reference [elliott2006binomial] (which lists many equivalent conditions), a
+binomial ring is a torsion-free commutative ring `R` such that for any `x ∈ R` and any `k ∈ ℕ`, the
+product `x(x-1)⋯(x-k+1)` is divisible by `k!`.  The torsion-free condition lets us divide by `k!`
+unambiguously, so we get uniquely defined binomial coefficients.
+
+The defining condition doesn't require commutativity, and we get a theory with essentially the same
+power by replacing subtraction with addition.  Thus, we consider a semiring `R` in which
+multiplication by factorials is injective, and demand that for all `x ∈ R` and any `k ∈ ℕ`,
+`x(x+1)⋯(x+(k-1))` is divisible by `k!`.
+The quotient is called `multichoose r k`, following the convention given for natural numbers.
+
+## References
+* [J. Elliott, *Binomial rings, integer-valued polynomials, and λ-rings*][elliott2006binomial]
+
+-/
+
+universe u
+
+variable {R : Type u}
+
+section Semiring
+
+variable [Semiring R]
+
+namespace Ring
+
+/-- `ascFactorial`, also known as the rising factorial function, is the product
+`r (r+1) ⋯ (r + n - 1)`.  In other words, it recursively evaluates the `n`-th ascending Pochhammer
+polynomial at a semiring element `r`.-/
+def ascFactorial (r : R) : ℕ → R
+  | 0 => 1
+  | (k + 1) => (ascFactorial r k) * (r + k)
```

Comment:
This is just `ascPochhammer` as a bare function instead of a polynomial. Why not use `(ascPochhammer n k).eval`?

Reply:
I had found it a bit cumbersome to work with the polynomials earlier, but that was when I was just starting Lean.  I'll try again.